src/main/java/graphql/schema/visitor/GraphQLSmartTypeVisitor.java

```diff
@@ -0,0 +1,52 @@
+package graphql.schema.visitor;
+
+import graphql.PublicSpi;
+import graphql.schema.GraphQLFieldDefinition;
+import graphql.schema.GraphQLFieldsContainer;
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.GraphQLTypeVisitor;
+import graphql.util.TraversalControl;
+
+/**
+ * I called it smart because I want to offer more "smarts" above GraphQLTypeVisitor and its "context" uber object
+ * <p>
+ * You would use it by doing `new GraphQLSmartTypeVisitor() { ...}.toTypeVisitor()`
+ */
+@PublicSpi
+public interface GraphQLSmartTypeVisitor {
+
+    default TraversalControl visitGraphQLObjectType(GraphQLObjectType objectType, SmartTypeVisitorEnvironment environment) {
+        return TraversalControl.CONTINUE;
+    }
+
+    /**
+     * This is a class specific for visiting {@link GraphQLFieldDefinition}s
+     */
+    interface GraphQLFieldDefinitionVisitorEnvironment extends SmartTypeVisitorEnvironment {
+
+        GraphQLFieldsContainer getFieldsContainer();
```

Comment:
One other option here is to NOT have a per method interface but rather put this into method signature

```
visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, GraphQLFieldsContainer 
 container, SmartTypeVisitorEnvironment environment) {
```

Reply:
What else would go here to make it smarter ?

Reply:
What about some callback methods for the intermediate interfaces in the class hierarchy?
Like, I often find myself having to do stuff on all elements that can have directives. 
Maybe a callback like `visitGraphQLDirectiveContainer(GraphQLDirectiveContainer directiveContainer.....)`? `GraphQLFieldsContainer` is another very common one.
Although that can quickly become convoluted - do we need one for each of those intermediate interfaces (`GraphQLNamedInputType`, `GraphQLNullableType`, etc...)? Probably not.

Reply:
maybe like a dynamic visitor method`<T extends GraphQLSchemaElement> visitThing(Class<T> thingClass, BiConsumer<T, Context>)` , and in the visitor we could check `thingClass.isAssignableFrom(element.getClass())`. Something like that...
Not sure if there's already something like that in the API, though.